### ISSUE TITLE
`finetune_gradient_checkpointing` has no effect on the transformers path

### ISSUE THREAD

**Issue by Derek-w-steinle** [sam-paech/auto-antislop#3]
Thanks for releasing this, I've been reproducing the paper's results with gemma-3-4b and gemma-3-12b and hit an OOM I think traces back to two things in the non-unsloth code path. Both look like straightforward fixes, and I have before/after numbers.

## 1. Gradient checkpointing is never enabled when `finetune_use_unsloth: false`

`finetune_gradient_checkpointing` is read in exactly one place, `core/finetuning.py:497`, which is inside the `if use_unsloth:` branch:

```python
model = FastLanguageModel.get_peft_model(
    ...
    use_gradient_checkpointing=config['finetune_gradient_checkpointing'],
```

The transformers/TRL branch (`core/finetuning.py:436-520`) never calls `gradient_checkpointing_enable()`, and the `ConfigClass(...)` at ~line 766 doesn't pass `gradient_checkpointing=True`. So on that path the setting is silently inert.

I ended up on the transformers path because installing unsloth pulls its own torch and breaks the pinned vllm/trl stack, which the README also warns about.

Suggested fix, right after `model = get_peft_model(model, lora_cfg)`:

```python
if config.get('finetune_gradient_checkpointing', True):
    model.gradient_checkpointing_enable(
        gradient_checkpointing_kwargs={'use_reentrant': True})
    model.enable_input_require_grads()
    model.config.use_cache = False
```

One note on `use_reentrant`. I tried `False` first, since it's the usual recommendation, and training died at step 0 on a checkpoint metadata mismatch: the saved tensor was `[1, 4000, 3840] float32` against a recomputed `[4000, 3840] bfloat16`. I take this to be the custom `position_ids` and the packed FTPO batch shape, but I didn't chase it further once `True` worked. Flagging it so nobody else loses a run to it.

## 2. The FTPO trainer materialises logits for every position but uses one

In `core/ftpo_trainer.py`, all three forward passes (two on the policy model at lines 218 and 260, one on `self.ref_model` at line 265) compute the full `[B, L, V]` logits and then slice the final position:

```python
outputs = model(ids, attention_mask=attn, position_ids=pos_full,
                use_cache=False, return_dict=True)
logits_last = outputs.logits[:, -1, :]   # [B, V]
```

Since `[:, -1, :]` is a view, the full tensor stays alive through the loss. For gemma-3-12b (V=262,208) at batch 1 x seq 4000 in bf16 that's ~1.95 GiB per forward, three forwards per step, before counting any float32 copies the loss math makes.

Passing `logits_to_keep=1` is mathematically identical here (the slice still selects the last position) and avoids building the rest:

```python
outputs = model(ids, attention_mask=attn, position_ids=pos_full,
                use_cache=False, return_dict=True, logits_to_keep=1)
```

Applies at lines 218, 260 and 265.

## Measurements

gemma-3-12b, LoRA rank 256, all 8 target modules, 10 unfrozen layers, 4-bit base, batch 1, grad accum 16, max seq 4000, eager attention, `finetune_use_unsloth: false` (config copied from the 4b yaml with model_id swapped):

| | Peak VRAM | Result |
|---|---|---|
| Before | 139.6 GiB | OOM on a 141GB H200 |
| Before, rank 128, 6 of 10 layers unfrozen | 138.9 GiB | OOM on the same card |
| After both fixes, rank 256, 10 layers | **28.3 GiB** | **converged on an 80GB A100** |

The middle row is what pointed at the cause. Cutting the adapters roughly in half, and the unfrozen layer count with them, moved the total by half a percent, so the dominant term wasn't coming from the adapters at all.

The patched run hit the early-stopping threshold at step 35 of 69 (`chosen_win` 0.8566, loss 3.9119 to 1.0184) and saved adapters normally, so the fixes don't appear to cost anything in training behaviour.

Fuller write-up with the training curve, and the patch as a standalone script, in my reproduction repo: https://github.com/Derek-w-steinle/antislop-reproduction

Happy to open a PR with both changes if useful.

**Comment by sam-paech** [comment]
Good finds, yes a pr would be great.
